File: 25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def f(head,tail):
            pre,nxt=None,head 
            while head!=tail:
                nxt = head.next 
                head.next = pre 
                pre=head
                head=nxt 
            return pre
        if not head or not head.next:
            return head
        tail=head  
        for i in range(k):
            if not tail:
                return head 
            tail = tail.next 
        last = f(head,tail)
        head.next = self.reverseKGroup(tail, k)
        return last 

        
# @lc code=end

File: 25.k-个一组翻转链表.java
/*
 * @lc app=leetcode.cn id=25 lang=java
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // public ListNode reverseKGroup(ListNode head, int k) {
    //     ListNode dummy = new ListNode(-1,head);
    //     ListNode pre=dummy,end=dummy,endNext=null;
        
    //     while(end.next!=null){
    //         for(int i=0;i<k && end!=null;++i)
    //             end=end.next;
    //         if(end==null)
    //             break;
    //         ListNode start = pre.next;
    //         endNext = end.next;
    //         end.next = null;
    //         pre.next = reverseALL(start);
    //         start.next = endNext;
    //         pre=start;
    //         end=start;
    //     }
    //     return dummy.next;
    // }
    // public ListNode reverseALL(ListNode head){
    //     ListNode cur=null,pre=null;
    //     while(head!=null){
    //         cur=head.next;
    //         head.next = pre;
    //         pre=head;
    //         head = cur;
    //     }
    //     return pre;
    // }
    public ListNode reverseKGroup(ListNode head, int k) { 
        ListNode dummy = new ListNode(-1,head),end=dummy,pre=dummy,endNext=null;
        while(end.next!=null){
            for(int i=0;i<k && end!=null;++i){
                end=end.next;
            }
            if(end==null) break;
            ListNode start=pre.next;
            endNext = end.next;
            end.next=null;
            pre.next = reverseAll(start);
            start.next = endNext;
            pre=start;
            end=start;
        }
        return dummy.next;
    }
    public ListNode reverseAll(ListNode head){
        ListNode pre = null,next=null;
        while(head!=null){
            next = head.next;
            head.next=pre;
            pre=head;
            head=next;
        }
        return pre;
    }
}
// @lc code=end

